promql/engine.go

```diff
@@ -1359,46 +1359,43 @@ func (ev *evaluator) rangeEvalAgg(ctx context.Context, aggExpr *parser.Aggregate
 	}
 	groups := make([]groupedAggregation, groupCount)
 
-	var k int64
-	var ratio float64
 	var seriess map[uint64]Series
+
 	switch aggExpr.Op {
 	case parser.TOPK, parser.BOTTOMK, parser.LIMITK:
-		if !convertibleToInt64(param) {
-			ev.errorf("Scalar value %v overflows int64", param)
-		}
-		k = int64(param)
-		if k > int64(len(inputMatrix)) {
-			k = int64(len(inputMatrix))
-		}
-		if k < 1 {
-			return nil, warnings
+		// Return early if all k values are less than one.
+		if params.Max() < 1 {
+			return nil, annos
 		}
-		seriess = make(map[uint64]Series, len(inputMatrix)) // Output series by series hash.
+		seriess = make(map[uint64]Series, len(inputMatrix))
+
 	case parser.LIMIT_RATIO:
-		if math.IsNaN(param) {
-			ev.errorf("Ratio value %v is NaN", param)
+		// Return early if all r values are zero.
+		if params.Max() == 0 && params.Min() == 0 {
+			return nil, annos
 		}
-		switch {
-		case param == 0:
-			return nil, warnings
-		case param < -1.0:
-			ratio = -1.0
-			warnings.Add(annotations.NewInvalidRatioWarning(param, ratio, aggExpr.Param.PositionRange()))
-		case param > 1.0:
-			ratio = 1.0
-			warnings.Add(annotations.NewInvalidRatioWarning(param, ratio, aggExpr.Param.PositionRange()))
-		default:
-			ratio = param
+		if maxVal := params.Max(); maxVal > 1.0 {
+			annos.Add(annotations.NewInvalidRatioWarning(maxVal, 1.0, aggExpr.Param.PositionRange()))
+		}
+		if minVal := params.Min(); minVal < -1.0 {
+			annos.Add(annotations.NewInvalidRatioWarning(minVal, -1.0, aggExpr.Param.PositionRange()))
 		}
-		seriess = make(map[uint64]Series, len(inputMatrix)) // Output series by series hash.
+		seriess = make(map[uint64]Series, len(inputMatrix))
+
 	case parser.QUANTILE:
-		if math.IsNaN(param) || param < 0 || param > 1 {
-			warnings.Add(annotations.NewInvalidQuantileWarning(param, aggExpr.Param.PositionRange()))
+		if params.HasAnyNaN() {
+			annos.Add(annotations.NewInvalidQuantileWarning(math.NaN(), aggExpr.Param.PositionRange()))
+		}
+		if maxVal := params.Max(); maxVal > 1 {
+			annos.Add(annotations.NewInvalidQuantileWarning(maxVal, aggExpr.Param.PositionRange()))
+		}
+		if minVal := params.Min(); minVal < 0 {
+			annos.Add(annotations.NewInvalidQuantileWarning(minVal, aggExpr.Param.PositionRange()))
```

Comment:
As above, I would go for brevity and just call `params.Max()` etc multiple times.